### ISSUE TITLE
Queue errors prior to initialisation

### ISSUE THREAD

**Issue by wheresrhys** [Financial-Times/o-errors#12]
The use case is that oErrors is instantiated within next-js-setup, but this in turn has a dependency on next-flags-client which logs information about flags in use. Currently calling any of the oErrors logger methods prior to initialisation throws an error

**Comment by samgiles** [comment]
Just to clarify: do you mean the errors or the context logging?  Errors are buffered if you use the reporting iinterfaces - the following test should cover thaglt use case:
https://github.com/Financial-Times/o-errors/blob/master/test/test_errors.js#L226

**Comment by wheresrhys** [comment]
Context logging I think. The .error, .info methods

On Thursday, May 28, 2015, Samuel Giles <notifications@github.com> wrote:

> Just to clarify: do you mean the errors or the context logging? Errors are
> buffered if you use the reporting iinterfaces - the following test should
> cover thaglt use case:
>
> https://github.com/Financial-Times/o-errors/blob/master/test/test_errors.js#L226
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Financial-Times/o-errors/issues/12#issuecomment-106550078>
> .
>

-- 

------------------------------
This email was sent by a company owned by Pearson plc, registered office at 
80 Strand, London WC2R 0RL.  Registered in England and Wales with company 
number 53723.

**Comment by samgiles** [comment]
:+1:

**Comment by samgiles** [comment]
@wheresrhys  It's hard to do context logging correctly before initialisation as the options initialise the logging behaviour: context log level, default size of logs to keep, whether to record logs at all.  

Only real errors are buffered *if* they are reported via the interface o-errors provides because Raven installs the handler on `window.onerror` which isn't initialised until `init` is called.  We could install a `window.onerror` handler I suppose to buffer there before initialisation.

So basically, it's a bit weird to buffer contextual logging if they're not going to be used at all, but we can do it.

**Comment by wheresrhys** [comment]
In that case maybe it should stay out of o-errors, but could it in that case expose whether it's initialised or not and fire an event on initialisation? Or a method such as

```javascript
queue: function (func) {
   if (I am initialised) {
    func();
  } else {
    this.queue.push(func);
  }
}

// within init()

queue.forEach(function (func) { func()});

```

**Comment by samgiles** [comment]
As discussed you get this when reporting via the 'report' API.  Rhys is that sufficient enough to close this issue?  I have a branch I'm slowly working  on to improve some of the API docs and link to the rendered JS Docs. // cc @wheresrhys

**Comment by wheresrhys** [comment]
yeah sure - cheers for your help
